src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,840 @@
+                        iterator.remove();
+                    }
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            try {
+                if (selector != null) {
+                    // close all connections
+                    for (SelectionKey keys : selector.keys()) {
+                        keys.channel().close();
+                    }
+                    selector.close();
+                }
+                if (serverSocketChannel != null) {
+                    serverSocketChannel.close();
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            log.info("neighbor router stopped");
+        }
+    }
+
+    private static LongHashFunction txCacheDigestHashFunc = LongHashFunction.xx();
+
+    /**
+     * Computes the digest of the given transaction data.
+     * 
+     * @param txBytes The raw byte encoded transaction data
+     * @return The the digest of the transaction data
+     */
+    public static long getTxCacheDigest(byte[] txBytes) {
+        return txCacheDigestHashFunc.hashBytes(txBytes);
+    }
+
+    /**
+     * Adds the given neighbor to the 'reconnect pool' of neighbors which this node will try to reconnect to. If the
+     * domain of the neighbor was known when the connection was established, it will be used to re-establish the
+     * connection to the neighbor, otherwise the neighbor's current known IP address is used.<br/>
+     * The neighbor is only added to the 'reconnect pool' if the neighbor was ready to send/process messages.
+     * 
+     * @param neighbor the neighbor to attempt to reconnect to
+     */
+    private void addToReconnectPool(Neighbor neighbor) {
+        if (neighbor.getState() != NeighborState.READY_FOR_MESSAGES) {
+            return;
+        }
+        // try to pull out the origin domain which was used to establish
+        // the connection with this neighbor
+        String domain = ipToDomainMapping.get(neighbor.getHostAddress());
+        URI reconnectURI = null;
+        if (domain != null) {
+            reconnectURI = URI
+                    .create(String.format("%s%s:%d", protocolPrefix, domain, neighbor.getRemoteServerSocketPort()));
+        } else {
+            reconnectURI = URI.create(String.format("%s%s", protocolPrefix, neighbor.getHostAddressAndPort()));
+        }
+        log.info("adding {} to the reconnect pool", reconnectURI);
+        reconnectPool.add(reconnectURI);
+    }
+
+    /**
+     * Ensures that the neighbor is removed from the reconnect pool by using the neighbor's IP address and domain
+     * identity.
+     * 
+     * @param neighbor the neighbor to remove from the reconnect pool
+     */
+    private void removeFromReconnectPool(Neighbor neighbor) {
+        URI raw = URI.create(String.format("%s%s", protocolPrefix, neighbor.getHostAddressAndPort()));
+        reconnectPool.remove(raw);
```

Comment:
It could be that someone adds the same neighbor via `addNeighbor` with the IP and with the domain. When then the entry with the domain wins the race to gain the connection/is handshaked, this ensures that the entries are cleaned from the reconnect pool.